---
title: "R Notebook"
output: html_notebook
---


# Calculate index switch rate

While using unique dual-indices will allow detection and removal of the majority of index switch reads, there will still be low level undetectable index switching present at a rate of obs/exp^2 (ref- ). to determine this rate, we will first calculate the unexpected index combinations compared to the expected. 

Fastq files contain the index information for each read in the read header, and therefore to get all undetermined indices, both switched and otherwise erroneous we can summarise the index sequences for each read as contained in the fasta header:

@M03633:307:000000000-D4262:1:1101:19524:28535 1:N:0:**GAGACGAT+GTTCTCGT**
ATACTGTGCGTACTGCAGATCGGAAGAGCACACGTCTGAACTCCAGTCACGAGACGATATCTCGTATGCCGTCTT 
+ 
BBBB?FFFBABAEGGGGGGGGGGGGGFHHHHHHGHHHGHHHHHHHHHHHHGGEEEEEAFGGHGHFAHHHHGGGH

First we need to demultiplex the data again allowing no mismatches - This needs to be done with bcl2fastq rather than the default illumina miseq lociratefastq workflow, as the workflow doesnt include indices in fastq file headers

BASH:
```{bash demultiplex 1 mismatch}
###BASH###

#raise amount of available file handles
ulimit -n 4000

#Miseq Run 1 - Testing Primers
/home/ap0y/usr/local/bin/bcl2fastq -p 12 --runfolder-dir /group/sequencing/190331_M03633_0310_000000000-CB3DR  --output-dir /group/pathogens/Alexp/Metabarcoding/dros_metabarcoding/data/run1_CB3DR --sample-sheet /group/pathogens/Alexp/Metabarcoding/dros_metabarcoding/data/run1_CB3DR/SampleSheet_CB3DR.csv --no-lane-splitting --barcode-mismatches 0

#Miseq Run 2 - Testing replicate tagged primers & SynMock
/home/ap0y/usr/local/bin/bcl2fastq -p 12 --runfolder-dir /group/sequencing/190628_M03633_0331_000000000-CK3HD  --output-dir /group/pathogens/Alexp/Metabarcoding/dros_metabarcoding/data/run2_CK3HD --sample-sheet /group/pathogens/Alexp/Metabarcoding/dros_metabarcoding/data/run2_CK3HD/SampleSheet_CK3HD.csv --no-lane-splitting --barcode-mismatches 0

#Miseq Run 3 - Ladder spike ins
/home/ap0y/usr/local/bin/bcl2fastq -p 12 --runfolder-dir /group/sequencing/190722_M03633_0336_000000000-CJKFJ/  --output-dir /group/pathogens/Alexp/Metabarcoding/dros_metabarcoding/data/run3_CJKFJ --sample-sheet /group/pathogens/Alexp/Metabarcoding/dros_metabarcoding/data/run3_CJKFJ/SampleSheet_CJKFJ.csv --no-lane-splitting --barcode-mismatches 0
```


Then we can summarise the switch rate by counting unassigned reads

First you need to change the grep call to specifically call the sequencers ID:

Agribio miseq 1:
Agribio miseq 2: @M03633
Agribio NovaSeq: @A00878

```{bash undetermined}
###BASH###
#Run 1
#summarise undetermined reads from R1 undetermined file
zcat Undetermined_S0_L001_R1_001.fastq.gz | grep '^@A00878' | cut -d : -f 10 | sort | uniq -c | sort -nr > undetermined.txt

# summarise correctly determined reads from all other R1 files 
rm determined.txt
ls | grep "R1_001.fastq.gz" | sort | grep -v 'Undetermined' > test_ls_F

let files=$(grep -c "fastq.gz" test_ls_F)

declare -i x

x=1
while [ $x -le $files ] 
    do

query=$(sed -n "${x}p" test_ls_F)

sample_name=$(echo $query | awk -F . '{ print $1}')

stats=$(zcat $(echo $query) | grep '^@A00878' | wc -l)
echo $query $stats >> determined.txt

let x=x+1

done 

```

To differentiate unused indices arising from switching, from unused indices arrising from other phenomena, we can compare the undetermined count file to all possible combinations of i5 and i7 indices that could be produced through switching 

Can i just bind the read numbers onto the samplesheet here? - can get from a HTML file in /group/sequencing/191015_A00878_0012_AHLVKYDMXX/Unaligned-Single8-Lane1/Reports/html/HLVKYDMXX/all/all/all


```{r index switching, eval=TRUE}
#Read in original sample sheet
SampleSheet <- read_csv("Run4_nova/SampleSheet_run4.csv",skip=20)

##For special case of popgen spikein
SampleSheet <- read_csv("Run3_spikein/SampleSheet_run3.csv",skip=20)  %>%
  filter(str_detect(Sample_Name,pattern="-"))

#Create all possible switched combinations
combos <- unique(expand.grid(SampleSheet$index, SampleSheet$index2))
combos$indices <- paste0(combos$Var1,"+",combos$Var2)

#Determined reads from mock communities
determined <- read_table2("Run4_nova/determined.txt",col_names = FALSE) %>%
  set_colnames(c("Sample_Name","count")) %>%
  mutate(Sample_Name = Sample_Name %>% str_split_fixed("_S",n=2) %>% as_tibble() %>% pull(V1)) %>%
  mutate(Sample_Name = str_replace(Sample_Name, pattern="HLVKYDMXX_", replacement = "")) %>% #remove FCID from novaseq
  left_join(SampleSheet, by="Sample_Name") %>%
  unite(indices,c("index","index2"), sep="+") %>%
  select(c("Sample_Name","count","indices"))

#Undetermined reads from mock communities
undetermined <- read_table2("Run4_nova/undetermined.txt",col_names = FALSE) %>%
  set_colnames(c("count","indices")) %>%
  mutate(Sample_Name = "Undetermined_S0_R1_001.fastq.gz")

head(undetermined)

indices <- rbind(determined,undetermined)

#Calculate total read count for run
total_reads <- sum(indices$count)

#get unused combinations resulting from index switching
switched <- left_join(combos,indices,by="indices") %>%
  drop_na()
colnames(switched) <- c("i7","i5","indices","Sample_Name","count")

#get unused combinations resulting from other phenomena
other <- indices[!indices$indices %in% combos$indices, ]

#Count number of other undetermined
other_reads <- sum(other$count)

##Summary of index switching rate
exp_rate <- switched %>% 
  filter(!str_detect(Sample_Name,"Undetermined"))
obs_rate <- switched %>% 
  filter(str_detect(Sample_Name,"Undetermined"))

switch_rate <- (sum(obs_rate$count)/sum(exp_rate$count))
message(switch_rate)

#Rate of undetected switching should be switch_rate squared
filt_threshold <- switch_rate^2
message(paste0("The threshold for filtering will be: ",filt_threshold))

#Plot switching

gg.switch <- ggplot(data = switched, aes(x = i7, y = i5), stat="identity") +
  geom_tile(aes(fill = count),alpha=0.8)  + scale_fill_viridis(name = "reads", begin=0.1,trans = "log10")  + 
  theme(axis.text.x = element_text(angle=90, hjust=1), plot.title=element_text(hjust = 0.5), plot.subtitle =element_text(hjust = 0.5), legend.position = "none") +
  labs(title= "Novaseq run", subtitle = paste0("Total Reads: ", total_reads, " Switch rate: ", sprintf("%1.2f%%", switch_rate*100)))


#Plot pooling

gg.pooling <- ggplot(data=determined, aes(x=fct_rev(Sample_Name),y=count),stat="identity") + 
  geom_bar(aes(fill=count),stat="identity")  + 
  scale_fill_viridis(name = "reads", begin=0.1) + 
  theme(axis.text.x = element_text(angle=90, hjust=1), plot.title=element_text(hjust = 0.5), plot.subtitle =element_text(hjust = 0.5))+ 
  geom_hline(aes(yintercept = mean(count)))  +
  xlab("Sample")+
  ylab("number of reads") +
  labs(title= "Pooling", subtitle = paste0("Total Reads: ", total_reads, " Average reads: ",  sprintf("%.0f",mean(determined$count))," Standard deviation: ", sprintf("%.0f",sd(determined$count)))) +
  coord_flip()


##Special case - plot pooling of popgen spiked

popgen <- determined %>%
  filter(!str_detect(Sample_Name,pattern="-"))

gg.popgen <- ggplot(data=popgen, aes(x=fct_rev(Sample_Name),y=count),stat="identity") + 
  geom_bar(aes(fill=count),stat="identity")  + 
  scale_fill_viridis(name = "reads", begin=0.1) + 
  theme(axis.text.x = element_text(angle=90, hjust=1), plot.title=element_text(hjust = 0.5), plot.subtitle =element_text(hjust = 0.5))+ 
  geom_hline(aes(yintercept = mean(count)))  +
  xlab("Synthetic sample")+
  ylab("number of reads") +
  #labs(title= "Pooling", subtitle = paste0("Total Reads: ", total_reads, " Average reads: ",  #sprintf("%.0f",mean(determined$count))," Standard deviation: ", sprintf("%.0f",sd(determined$count)))) +
  coord_flip()


```

