# iMapPESTS

This repository contains an R based workflow based around the DADA2 algorithm for analysing metabarcoding data generated through the iMapPESTS project.

There are 2 seperate workflows and vignettes available:
* [Local version for analysis on windows/linux PC](https://alexpiper.github.io/iMapPESTS/local_metabarcoding.html)
* Slurm cluster version for analysis on Agriculture Victoria BASC server **COMING SOON**

The main workflow can be followed up with the following comparisons depending on what you are interested in
* [Diversity analyses](https://alexpiper.github.io/iMapPESTS/diversity.html)

Before starting, i recommend learning about the individual componenets of the workflow:
* [The DADA2 algorithm](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4927377/)
* [The benefits of amplicon sequence variants over OTU's](https://www.nature.com/articles/ismej2017119)
* [Taxonomic assignment using IDTAXA](https://microbiomejournal.biomedcentral.com/articles/10.1186/s40168-018-0521-5)

Some other helpful tutorials include:

* [DADA2 pipeline official tutorial](https://benjjneb.github.io/dada2/tutorial.html) By Benjamin Callahan.
* [DADA2 example workflow](https://astrobiomike.github.io/amplicon/dada2_workflow_ex) by Mike Lee.
